lib/shopify_transporter/pipeline/magento/order/top_level_attributes.rb

```diff
@@ -25,13 +34,112 @@ def convert(hash, record)
+
+          def total_paid(hash)
+            return hash['total_paid'].to_i if hash['total_paid'].present?
+            0
+          end
+
+          def total_refunded(hash)
+            return hash['total_refunded'].to_i if hash['total_refunded'].present?
+            0
+          end
+
+          def paid?(hash)
+            total_price(hash) == total_paid(hash) && total_refunded(hash) == 0
+          end
+
+          def partially_paid?(hash)
+            total_paid(hash) > 0 && total_paid(hash) < total_price(hash) && total_refunded(hash) == 0
+          end
+
+          def partially_refunded?(hash)
+            total_refunded(hash) > 0 && total_refunded(hash) < total_paid(hash)
+          end
+
+          def refunded?(hash)
+            total_refunded(hash) == total_paid(hash) && total_refunded(hash) != 0
+          end
+
+          def financial_status(hash)
+            order_state = hash['state']
+            status = nil
+
+            if order_state == 'Pending Payment'
+              status = 'pending'
+            elsif paid?(hash)
+              status = 'paid'
+            elsif partially_paid?(hash)
+              status = 'partially_paid'
+            elsif partially_refunded?(hash)
+              status = 'partially_refunded'
+            elsif refunded?(hash)
+              status = 'refunded'
+            end
+            status
+          end
+
+          def total_qty_shipped(hash)
+            return 0 unless hash.dig('items', 'result', 'items', 'item').present?
+            line_items = hash['items']['result']['items']['item']
+            return line_items['qty_shipped'].to_i if line_items.is_a?(Hash)
+            line_items.map { |line_item| line_item['qty_shipped'].to_i }.sum
+          end
+
+          def fulfillment_status(hash)
+            total_qty_ordered = hash['total_qty_ordered'].to_i
+            total_qty_shipped = total_qty_shipped(hash)
+            status = nil
+
+            if total_qty_shipped == total_qty_ordered
+              status = 'fulfilled'
+            elsif total_qty_shipped > 0 && total_qty_shipped < total_qty_ordered
+              status = 'partial'
+            end
+            status
+          end
+
+          def cancelled?(hash)
+            hash['state'] == 'canceled'
+          end
+
+          def cancelled_at(hash)
+            timestamp(hash, 'canceled') if cancelled?(hash)
+          end
+
+          def closed?(hash)
+            hash['state'] == 'closed'
+          end
+
+          def closed_at(hash)
+            timestamp(hash, 'closed') if closed?(hash)
+          end
+
+          def timestamp(hash, status)
+            return unless hash.dig('items', 'result', 'status_history', 'item').present?
+            timestamp = nil
+
+            history_items = hash['items']['result']['status_history']['item']
+            return history_items['created_at'] if history_items.is_a?(Hash) && history_items['status'] == status
+
+            history_items.each do |status_history|
```

Comment:
Minor: Is the intent here to select the last `history_item` that matches the status?

How do you feel about selecting the last item and then setting this?

```
last_applicable_history_item = history_items.select { |history_item| history_item['status'] == status }.last
timestamp = last_applicable_history_item['created_at'] unless last_applicable_history_item.nil?
```

The benefit of this is just doing the assignment once, but because we don't expect there to be a lot of history items, I think the current code is fine/feel free to leave as is.

Reply:
Good point! I've updated the code.